smart_open/bytebuffer.py

```diff
@@ -0,0 +1,74 @@
+# -*- coding: utf-8 -*-
+"""Implements a byte buffer that allows callers to read data with minimal
+copying, and has a fast __len__ method. The buffer is parametrized by its
+chunk_size, which is the number of bytes that it will read in from the supplied
+reader or iterable when the buffer is being filled. As primary use case for this
+buffer is to amortize the overhead costs of transferring data over the network
+(rather than capping memory consumption), it leads to more predictable
+performance to always read the same amount of bytes each time the buffer is
+filled, hence the chunk_size parameter instead of some fixed capacity.
+
+The bytes are stored in a bytestring, and previously-read bytes are freed when
+the buffer is next filled (by slicing the bytestring into a smaller copy)."""
+
+class ByteBuffer(object):
+    def __init__(self, chunk_size):
+        """The chunk_size indicates the number of bytes that will be read from
+        the supplied reader when filling the buffer. The buffer has no maximum
+        size."""
+        self._chunk_size = chunk_size
+        self.empty()
+
+    def __len__(self):
+        """Return the number of unread bytes in the buffer"""
+        return len(self._bytes) - self._pos
+
+    def read(self, size = -1):
```

Comment:
[PEP8](https://www.python.org/dev/peps/pep-0008/), here and everywhere.

Please run flake8 on your code and fix any errors that it reports.

```suggestion
    def read(self, size=-1):
```

Reply:
Fixed in https://github.com/RaRe-Technologies/smart_open/pull/284/commits/164339a2a836fb721578180a853d762bf67c3723